src/lib.rs

```diff
@@ -1638,34 +1654,53 @@ impl Bindings {
             cmd.args(&["--config-path", path]);
         }
 
-        if let Ok(output) = cmd.arg(file).output() {
-            if !output.status.success() {
-                let stderr = String::from_utf8_lossy(&output.stderr);
-                match output.status.code() {
-                    Some(2) => Err(io::Error::new(
-                        io::ErrorKind::Other,
-                        format!("Rustfmt parsing errors:\n{}", stderr),
-                    )),
-                    Some(3) => {
-                        warn!(
-                            "Rustfmt could not format some lines:\n{}",
-                            stderr
-                        );
-                        Ok(())
+        if let Ok(mut child) = cmd.spawn() {
+
+            let mut child_stdin = child.stdin.take().unwrap();
+            let mut child_stdout = child.stdout.take().unwrap();
+
+            // Write to stdin in a new thread, so that we can read from stdout on this
+            // thread. This keeps the child from blocking on writing to its stdout which
+            // might block us from writing to its stdin.
+            let stdin_handle = ::std::thread::spawn(move || {
+                let _ = child_stdin.write_all(source.as_bytes());
+
+                source
+            });
+
+            let mut output = vec![];
+            io::copy(&mut child_stdout, &mut output)?;
+
+            let status = child.wait()?;
+
+            let source = stdin_handle.join().unwrap();
```

Comment:
@fitzgen How (and should) we avoid the unwrap here?

Reply:
If we want to, we can 

```rust
.map_err(|_| io::Error::new(io::ErrorKind::Other, "panicked writing to `rustfmt` stdin".into())
```

and then use the `?` operator.

But I don't see any way that thread should be able to panic, so `unwrap` is probably fine.